File: lib/tools/linkedin-ads.ts
import { getCached, setCache } from '../supabase';
import type { ToolResult } from './types';
import { scrapePage } from './firecrawl';

export interface LinkedInAd {
  advertiser: string;
  adCopy: string;
  url: string;
}

export async function scrapeLinkedInAds(
  companyName: string
): Promise<ToolResult<LinkedInAd[]>> {
  const cacheKey = `linkedin-ads:${companyName}`;
  const cached = await getCached('linkedin_ads', cacheKey);
  if (cached) return { ...(cached as ToolResult<LinkedInAd[]>), cached: true };

  const searchUrl = `https://www.linkedin.com/ad-library/search?keywords=${encodeURIComponent(companyName)}`;

  try {
    const scraped = await scrapePage(searchUrl);
    const markdown = scraped.data.markdown ?? '';

    const ads: LinkedInAd[] = markdown
      .split('\n')
      .map(l => l.trim())
      .filter(l => l.length > 40 && !l.startsWith('#') && !l.startsWith('http'))
      .slice(0, 10)
      .map(line => ({ advertiser: companyName, adCopy: line, url: searchUrl }));

    const result: ToolResult<LinkedInAd[]> = {
      data: ads,
      source: 'LinkedIn Ad Library',
      sourceUrl: searchUrl,
      timestamp: new Date().toISOString(),
      confidence: ads.length > 0 ? 0.7 : 0.3,
      cached: false,
    };

    await setCache('linkedin_ads', cacheKey, result);
    return result;
  } catch {
    return {
      data: [],
      source: 'LinkedIn Ad Library (failed)',
      sourceUrl: searchUrl,
      timestamp: new Date().toISOString(),
      confidence: 0,
      cached: false,
    };
  }
}

export async function scrapeCompetitorLinkedInAds(
  competitors: string[]
): Promise<ToolResult<LinkedInAd[]>[]> {
  return Promise.all(competitors.map(c => scrapeLinkedInAds(c)));
}

File: backend/app/tools/linkedin_ads.py
"""LinkedIn Ad Library scrape — mirrors lib/tools/linkedin-ads.ts."""

from __future__ import annotations

from typing import Any
from urllib.parse import quote

from app.supabase_client import get_cached, set_cache

from .fallback import build_tool_result
from .firecrawl import scrape_page


async def scrape_linkedin_ads(company_name: str) -> dict[str, Any]:
    cache_key = f"linkedin-ads:{company_name}"
    cached = await get_cached("linkedin_ads", cache_key)
    if cached:
        return {**cached, "cached": True}

    search_url = (
        f"https://www.linkedin.com/ad-library/search?keywords={quote(company_name)}"
    )

    try:
        scraped = await scrape_page(search_url)
        markdown = (scraped.get("data") or {}).get("markdown") or ""

        ads = [
            {
                "advertiser": company_name,
                "adCopy": line,
                "url": search_url,
            }
            for line in (
                ln.strip()
                for ln in markdown.split("\n")
            )
            if len(line) > 40 and not line.startswith("#") and not line.startswith("http")
        ][:10]

        result = build_tool_result(
            data=ads,
            status="ok" if ads else "degraded",
            source="LinkedIn Ad Library",
            source_url=search_url,
            confidence_override=0.7 if ads else 0.3,
        )
        await set_cache("linkedin_ads", cache_key, result)
        return result
    except Exception:
        return build_tool_result(
            data=[],
            status="failed",
            source="LinkedIn Ad Library (failed)",
            source_url=search_url,
            confidence_override=0.0,
        )


async def scrape_competitor_linkedin_ads(
    competitors: list[str],
) -> list[dict[str, Any]]:
    import asyncio

    return list(await asyncio.gather(*(scrape_linkedin_ads(c) for c in competitors)))
